common/src/chain/config/emission_schedule.rs

```diff
@@ -0,0 +1,336 @@
+use crate::primitives::{Amount, BlockHeight};
+use std::time::Duration;
+
+// TODO Move to a separate module
+/// Represents a certain amount of MLT.
+#[derive(Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Debug)]
+pub struct Mlt(Amount);
+
+impl Mlt {
+    /// Number of decimal digits used to represent MLT
+    pub const DECIMALS: u8 = 11;
+    /// Number of atoms in 1 MLT
+    pub const ATOMS_PER_MLT: u128 = 10u128.pow(Self::DECIMALS as u32);
+    /// Zero MLTs
+    pub const ZERO: Self = Self(Amount::from_atoms(0));
+    /// Maximum representable amount of MLTs
+    pub const MAX: Self = Self(Amount::MAX);
+
+    /// Construct from the number atomic units
+    pub const fn from_atoms(n: u128) -> Self {
+        Self(Amount::from_atoms(n))
+    }
+
+    /// Construct from the number of MLTs
+    pub const fn from_mlt(n: u64) -> Self {
+        // Since the argument is u64 and number of atoms in 1 MLT is <= u64::MAX,
+        // the result is guaranteed to fit into the internal representation of Amount (u128)
+        static_assertions::const_assert!(Mlt::ATOMS_PER_MLT <= u64::MAX as u128);
+        Self(Amount::from_atoms(n as u128 * Mlt::ATOMS_PER_MLT))
+    }
+
+    /// Convert the number of atoms to Amount
+    pub const fn to_amount_atoms(self) -> Amount {
+        self.0
+    }
+
+    /// Multiplication, saturated at [Mlt::MAX]
+    pub fn saturating_mul(self, rhs: u128) -> Self {
+        Self::from_atoms(self.0.into_atoms().saturating_mul(rhs))
+    }
+
+    /// Addition, saturated at [Mlt::MAX]
+    pub fn saturating_add(self, rhs: Self) -> Self {
+        Self::from_atoms(self.0.into_atoms().saturating_add(rhs.0.into_atoms()))
+    }
+}
+
+impl std::ops::Add for Mlt {
+    type Output = Option<Self>;
+    fn add(self, rhs: Self) -> Option<Self> {
+        (self.0 + rhs.0).map(Self)
+    }
+}
+
+impl std::ops::Sub for Mlt {
+    type Output = Option<Self>;
+    fn sub(self, rhs: Self) -> Option<Self> {
+        (self.0 - rhs.0).map(Self)
+    }
+}
+
+impl std::ops::Mul<u128> for Mlt {
+    type Output = Option<Self>;
+    fn mul(self, rhs: u128) -> Option<Self> {
+        (self.0 * rhs).map(Self)
+    }
+}
+
+/// Internal emission schedule representation
+pub type EmissionScheduleFn = dyn Fn(BlockHeight) -> Mlt + Sync + Send + 'static;
+
+/// Emission schedule, characterized by function from block height to total supply at that point.
+///
+/// The function has to be a monotonic non-decreasing function. The function also has to
+/// eventually flatten out to a constant function if [Self::final_supply] is required to make any
+/// sense. The point at which the function becomes constant is where the block rewards run out.
+#[derive(Clone)]
+pub struct EmissionSchedule(std::sync::Arc<EmissionScheduleFn>);
```

Comment:
If you want to keep this as a function, that's OK. Just please present a way to read this from a string so that it can be passed using functional tests.

Reply:
I think the easy solution to this is to just make another caller for `from_yearly_table` that accepts somekind of string input and parses it.

Reply:
Ok, will add a string representation tomorrow.

Reply:
Thinking about this once again, there is a slight inconvenience in doing it at the level of `EmissionSchedule`. It is because `from_yearly_table` also takes the time interval between blocks as an argument, which is used also in other places. That means the user would have to specify the block interval in multiple places if the emission schedule was supposed to be constructed entirely from one string.

I think it would be an improvement to specify the reward drop period in number of blocks rather than making it yearly. I quite like this one since it does not introduce the notion of wall clock time into emission schedule, which otherwise works exclusively on blocks. Also allows user to specify different reward period, not just yearly. Conversion from year to number of blocks would be done elsewhere (or using a helper method here).

All in all the `from_yearly_table` (or equivalent) method requires three pieces of information:

1. The initial MLT supply
2. Per-block reward in each reward period (by default number of blocks corresponding to 1 year)
3. The reward drop period in the number of blocks (either passed in directly or calculated from block interval and period time)

When parsing these from the command line, I think these should be separate command line arguments. The reward table (2) could be a comma-separated list of decimal numbers. However due to way the drop period is linked with the block interval, I think it may best be done outside of this module.

My proposal is:
1. Add some tools to support table parsing to this PR:
  a. Change the reward block period to block count
  b. Add parsing of the per-period reward table from string (but not the whole emission schedule)
2. Integrate these things together in the subsequent PR (ChainConfig builder refactor):
  a. Can calculate number of blocks per year because ChainConfig has access to block interval
  b. Let user override the reward drop period (as block count or real time)
  c. Let user override the initial supply
3. In yet another PR
  a. Add the ability to build ChainConfig from the command line by exposing most of its methods

Cmments?

Reply:
> I'm not super-bullish on the `from_yearly_table`. It's only valuable because in Mintlayer we have yearly schedule for the coins on mainnet. When we use this function we're, by definition, not gonna use it for anything else that doesn't conform with the yearly schedule. The format I prefer is (as you saw before) a list of pair of reward/block-height in the form of "this is the reward from now on".

Ok. I think the representation deserves some thought so we can enforce as many of the invariants as possible by construction. That is monotonocity, totality and finiteness of supply. I think it is possible with minor modifications.

> There's no simpler format for this without making assumptions (the drop from initial reward works too, but it's O(N) in finding the current value and I feel it just complicates things).

If you mean starting with some initial reward and going down with the reward by a constant amount each block, that becomes a quadratic formula for total supply at given block height. That would be a pretty elegant supply schedule IMHO. I have a feeling that it's not what you had in mind though.

> The reason why I oppose the idea of making a function for the yearly assumption is that the tests will kind of become useless. I want everything to go through a single path and be tested from there by only varying the block height vs reward.
> 
> Right now I feel like we're dancing around the `from_yearly_table` function or something with a fixed period... I don't think we need to do this.

Ok will come up with something more general. There is the function representation that is about as general as it gets but that is not easily passed from command line as a string (unless we want to embed some sort of interpreter).

> I'm not opposed to the refactoring you did in general. But the yearly/periodic thing looks like a bad idea. Besides of the generic format I suggested, any function that will calculate the schedule and validate it should be generic enough to work for any kind of schedule, whether it's the one from mainnet or an arbitrary one coming from functional tests.

Well you could still set the period to one block and provide the reward for each block as a huge table explicitly but that's not very elegant.

> Basically, the required condition for this PR to look complete is just to be able to provide the block rewards arbitrarily as a string (whether that string function exists or not). I'm afraid with this PR, as is, we're cutting that off and accepting assumptions that aren't always true.